File: coding/src/com/algos/core/li15_dp/revised/OneD_Subseq_L213_HouseRobberII.java
package  com.algos.core.li15_dp.revised;

import java.util.Arrays;

public class OneD_Subseq_L213_HouseRobberII {

	public static void main(String[] args) {
		OneD_Subseq_L213_HouseRobberII obj = new OneD_Subseq_L213_HouseRobberII();
		
		int[] nums = new int[] {2,3,2};
		//int[] nums = new int[] {0,0};
		System.out.println(obj.rob(nums));
	}
	
	// Problem: https://leetcode.com/problems/house-robber-ii/
	// Idea: We can't rob both first and last houses. Hence we slip array in to 2 parts. 
	// One including the first element and another one including last element. 
	// Run normal House Robber algo on both arrays and take max.
	// Time: O(n)
	// Space: O(n)
	public int rob(int[] nums) {
        int n = nums.length;
        if(n == 1)
            return nums[0];
        
        int[] part1 = Arrays.copyOfRange(nums, 0, n-1);
        int[] part2 = Arrays.copyOfRange(nums, 1, n);
        return Math.max(robHelper(part1), robHelper(part2));
    }

	private int robHelper(int[] profit) {
		int n = profit.length;
        int dp[]=new int[n];
    	Arrays.fill(dp,-1);

		dp[0]= profit[0];
		for(int i=1 ;i<n; i++){
			int nonPick = 0 + dp[i-1];
			int pick = profit[i];
			if(i>1)
				pick += dp[i-2];
			
			dp[i]= Math.max(pick, nonPick);
		}
		return dp[n-1];
    }

}

File: dsa/topics/15_dp/python/one_d_subseq_l213_house_robber_ii.py
# DP / Subsequences
#
# Problem: https://leetcode.com/problems/house-robber-ii/
# Idea: We can't rob both first and last houses. Hence we slip array in to 2 parts.
# One including the first element and another one including last element.
# Run normal House Robber algo on both arrays and take max.
#
# Time: O(n)
# Space: O(n)

def rob(nums):
    n = len(nums)

    if n == 1:
        return nums[0]

    # Case 1: include first house, exclude last
    part1 = nums[0:n-1]

    # Case 2: exclude first house, include last
    part2 = nums[1:n]

    return max(robHelper(part1), robHelper(part2))


def robHelper(profit):
    n = len(profit)
    dp = [-1] * n

    dp[0] = profit[0]
    for i in range(1, n):
        nonPick = dp[i - 1]
        pick = profit[i]
        if i > 1:
            pick += dp[i - 2]

        dp[i] = max(pick, nonPick)

    return dp[n - 1]


# -------- Test --------
nums = [2, 3, 2]
print(rob(nums))  # Expected: 3

# nums = [0, 0]
# print(rob(nums))  # Expected: 0
